src/LayerEditor/ui/diagram_editor/diagram_scene.py

```diff
@@ -129,24 +142,40 @@ def new_point(self, pos, gitem, close = False):
         new_g_point = self.add_point(pos, gitem)
         if self.last_point is not None:
             self.add_segment(self.last_point, new_g_point)
+
         if not close:
             self.last_point = new_g_point
             pt = new_g_point.pos()
             self.move_aux_segment(pt, origin=pt)
         else:
             self.last_point = None
             self.hide_aux_line()
+        return self.decomposition.last_split_shapes
 
     def mouse_create_event(self, event):
         #transform = self.parent().transform()
         #below_item = self.itemAt(event.scenePos(), transform)
         below_item = self.below_item(event.scenePos())
         close = event.modifiers() & Qt.ControlModifier
-        self.new_point(event.scenePos(), below_item, close)
-        event.accept()
-
-        self.selection._selected.clear()
-        self.update_scene()
+        with better_undo.group("New point"):
+            split_items = self.new_point(event.scenePos(), below_item, close)
+            event.accept()
+
+            self.selection._selected.clear()
+            self.update_scene()
+            # update scene so new shapes would habe g_item for next step
+            for item in split_items:
```

Comment:
This should be done at the model layer. Probably during new_point call.

Reply:
We should possibly make own wrapper around decomposition to implement undoable operations and serialization / deserialization.

Reply:
Moving this part anywhere else would only complicate the code. The only solution would be to make the wrapper around decomposition as you suggested but I fear it would also result in more complicated code.

Reply:
Possibly just have separate file with bunch of functions to perform more complex data operations on the Decomposition.